src/strategy/filesystem.rs

```diff
@@ -0,0 +1,180 @@
+        anyhow::bail!("incorrect file permissions");
+    }
+
+    if attr.uid() != 0 {
+        anyhow::bail!("file not owned by root user");
+    }
+
+    Ok(true)
+}
+
+/// Check whether the finalization marker file has expired, if `notAfter` key
+/// exists.
+async fn is_expired(path: &str) -> Result<bool> {
+    match parse_expiry_timestamp(path).await? {
+        Some(expiry_timestamp) => {
+            // We can `unwrap()` since we're certain `UNIX_EPOCH` is in the past.
+            let current_timestamp = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs();
+            if current_timestamp > expiry_timestamp {
+                Ok(true)
+            } else {
+                Ok(false)
+            }
+        }
+        None => Ok(false),
+    }
+}
+
+#[context("failed to parse expiry timestamp from marker file")]
+async fn parse_expiry_timestamp(path: &str) -> Result<Option<u64>> {
+    let data = tokio_fs::read_to_string(path).await?;
+    let marker_json: FinalizationMarker =
+        serde_json::from_str(&data).context("failed to parse JSON content")?;
+    Ok(marker_json.not_after)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::fs;
+    use std::os::unix::fs::OpenOptionsExt;
+    use tempfile::tempdir;
+    use tokio::runtime as rt;
+
+    #[test]
+    fn test_bad_permissions_cannot_finalize() {
+        let tmp_dir = tempdir().unwrap();
+        let test_marker_file_path = tmp_dir.path().join("allowfinalize.json");
+        let test_marker_file_path = test_marker_file_path.to_str().unwrap();
+        fs::OpenOptions::new()
+            .create(true)
+            .write(true)
+            .mode(0o777)
+            .open(test_marker_file_path)
+            .unwrap();
+
+        let can_finalize = StrategyFilesystem::filesystem_allow_finalization(test_marker_file_path);
+        let runtime = rt::Runtime::new().unwrap();
+        runtime
+            .block_on(can_finalize)
+            .expect_err("file with incorrect permissions unexpectedly allowed finalization");
+    }
+
+    #[test]
+    fn test_parse_finalization_marker() {
+        let tmp_dir = tempdir().unwrap();
+        let test_marker_file_path = tmp_dir.path().join("allowfinalize.json");
+        let test_marker_file_path = test_marker_file_path.to_str().unwrap();
+        // 1619640863 is Apr 28 2021 20:14:23 UTC.
+        // Expect this to be expired.
+        let json = "
+        {
+            \"notAfter\": 1619640863
+        }";
+        fs::write(test_marker_file_path, json).expect("could not write test finalization file");
+        let expired = is_expired(test_marker_file_path);
+        let runtime = rt::Runtime::new().unwrap();
+        let expired = runtime.block_on(expired).unwrap();
+        assert_eq!(expired, true);
+
+        // Expect timepstamp with value `u64::MAX` to not be expired.
+        let json = serde_json::to_string(&json!({ "notAfter": u64::MAX })).unwrap();
```

Comment:
Ah I see you are using `json!` here, why not above?

Reply:
I used it here because I wanted to use `u64::MAX` more easily. But yeah I think I should've used it above too! :)